### PR TITLE
Implementation of StorageHandler

### PR DESCRIPTION
**Pull request by FirestarJes**
## Description

Implementation of storagehandler, to handle communication with the PostOffice storage.

### CODE REVIEW

File: source/Energinet.DataHub.PostOffice.Infrastructure/Services/MarketOperatorDataStorageService.cs [hunk 1]
```diff
@@ -14,28 +14,28 @@
...
         public async Task<Stream> GetMarketOperatorDataAsync(Uuid bundleUuid, Uri contentPath)
         {
             if (contentPath is null)
                 throw new ArgumentNullException(nameof(contentPath));
 
-            var container = new BlobContainerClient(_connectionString, _containerName);
-            var blob = container.GetBlobClient(contentPath.Segments.Last());
-            var response = await blob.DownloadStreamingAsync().ConfigureAwait(false);
-            return response.Value.Content;
+            return await _storageHandler.GetStreamFromStorageAsync(contentPath).ConfigureAwait(false);
```
Comment (michaelmajgaard) [inline source/Energinet.DataHub.PostOffice.Infrastructure/Services/MarketOperatorDataStorageService.cs:38]: You can skip `await` here, as the result is not used in scope. Just return the `Task<Stream>` directly and remove `async`, `await` and `.ConfigureAwait(false)`

File: source/PostOffice.Communicator.Tests/Storage/MockedBlobDownloadStreamingResult.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// // Unless required by applicable law or agreed to in writing, software
+// // distributed under the License is distributed on an "AS IS" BASIS,
+// // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// // See the License for the specific language governing permissions and
+// // limitations under the License.
+
+using System;
+using System.IO;
+using System.Reflection;
+using Azure.Messaging.ServiceBus;
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator.Tests/Storage/MockedBlobDownloadStreamingResult.cs:18]: Dont' think `Azure.Messaging.ServiceBus` is used here

File: source/Energinet.DataHub.PostOffice.Infrastructure/Services/MarketOperatorDataStorageService.cs (hunk 1 above)
Comment (FirestarJes) [reply source/Energinet.DataHub.PostOffice.Infrastructure/Services/MarketOperatorDataStorageService.cs]: Fixed

File: source/PostOffice.Communicator.Tests/Storage/MockedBlobDownloadStreamingResult.cs (hunk 2 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator.Tests/Storage/MockedBlobDownloadStreamingResult.cs]: Fixed

File: samples/SubDomain/GetMessage/Program.cs [hunk 3]
```diff
@@ -48,12 +49,15 @@ public static void Main()
                         new ServiceBusClient(serviceBusConnectionString));
                     services.AddSingleton<BlobServiceClient>(_ =>
                         new BlobServiceClient(blobStorageConnectionString));
+                    services.AddSingleton<IStorageHandler>(
+                        new StorageHandler(new StorageServiceClientFactory(blobStorageConnectionString)));
```
Comment (defectiveAi) [inline samples/SubDomain/GetMessage/Program.cs:53]: Jeg tror ikke du skal new StorageServiceClientFactory her, fordi du newer og registrerer den lidt længere nede.

File: samples/SubDomain/DataAvailable/DataAvailableNotification.csproj [hunk 4]
```diff
@@ -21,7 +21,7 @@ limitations under the License.
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="GreenEnergyHub.PostOffice.Communicator" Version="1.0.0-CI-04" />
+    <PackageReference Include="GreenEnergyHub.PostOffice.Communicator" Version="1.0.0-CI-05" />
```
Comment (defectiveAi) [inline samples/SubDomain/DataAvailable/DataAvailableNotification.csproj:24]: I see you are using v6 further down.

Comment (FirestarJes) [reply samples/SubDomain/DataAvailable/DataAvailableNotification.csproj:24]: Fixed

File: source/PostOffice.Communicator.SimpleInjector/source/PostOffice.Communicator.SimpleInjector/ContainerExtensions.cs [hunk 5]
```diff
@@ -63,5 +65,28 @@ private static void AddApplicationServices(this Container container)
...
         }
+
+        private static void AddStorageHandler(this Container container, string storageServiceConnectionStringConfigKey)
+        {
+            if (container == null)
+                throw new ArgumentNullException(nameof(container));
+
+            container.RegisterSingleton<IStorageServiceClientFactory>(() =>
+            {
+                var configuration = container.GetService<IConfiguration>();
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator.SimpleInjector/source/PostOffice.Communicator.SimpleInjector/ContainerExtensions.cs:76]: Extract 76..78 to a private static helper function, that takes container and key as parameter, and use it in both `AddStorageHandler` and `AddServiceBus`

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    {
+        [Fact]
+        public async Task AddStreamToStorageAsync_StreamIsEmpty_ThrowsArgumentException()
+        {
+            // arrange
+            var mockedStorageServiceClientFactory = new Mock<IStorageServiceClientFactory>();
+            var mockedBlobServiceClient = new Mock<BlobServiceClient>();
+            var mockedDataBundleRequestDto = new DataBundleRequestDto(
+                NewGuid().ToString(),
+                new List<Guid>() { NewGuid(), NewGuid(), NewGuid() });
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:44]: redundant `()`

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+                    x => x.Create())
+                .Returns(mockedBlobServiceClient.Object);
+
+            var target = new StorageHandler(mockedStorageServiceClientFactory.Object);
+
+            // act, assert
+            await Assert.ThrowsAsync<ArgumentNullException>(
+                    () => target.AddStreamToStorageAsync(
+                        Stream.Null,
+                        null))
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:77]: `null!` to tell resharper it is intented

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        public async Task AddStreamToStorageAsync_StreamIsValid_ReturnsUri()
+        {
+            // arrange
+            var mockedStorageServiceClientFactory = new Mock<IStorageServiceClientFactory>();
+            var mockedBlobServiceClient = new Mock<BlobServiceClient>();
+            var mockedBlobContainerClient = new Mock<BlobContainerClient>();
+            var mockedBlobClient = new Mock<BlobClient>();
+            var mockedDataBundleRequestDto = new DataBundleRequestDto(
+                NewGuid().ToString(),
+                new List<Guid>() { NewGuid(), NewGuid(), NewGuid() });
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:128]: redundant `()`

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        [Fact]
+        public async Task AddStreamToStorageAsync_RequestFailure_ThrowsCustomException()
+        {
+            // arrange
+            var mockedStorageServiceClientFactory = new Mock<IStorageServiceClientFactory>();
+            var mockedBlobServiceClient = new Mock<BlobServiceClient>();
+            var mockedBlobContainerClient = new Mock<BlobContainerClient>();
+            var mockedDataBundleRequestDto = new DataBundleRequestDto(
+                NewGuid().ToString(),
+                new List<Guid>() { NewGuid(), NewGuid(), NewGuid() });
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:90]: redundant `()`

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        [Fact]
+        public async Task GetStreamFromStorageAsync_ContentPathIsNull_ThrowsArgumentNullException()
+        {
+            // arrange
+            var mockedStorageServiceClientFactory = new Mock<IStorageServiceClientFactory>();
+            var target = new StorageHandler(mockedStorageServiceClientFactory.Object);
+
+            // act, assert
+            await Assert.ThrowsAsync<ArgumentNullException>(
+                    () => target.GetStreamFromStorageAsync(null))
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:215]: `null!` to tell resharper it is intented

File: source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Azure.Storage.Blobs;
+
+namespace GreenEnergyHub.PostOffice.Communicator.Factories
+{
+    /// <summary>
+    /// Factory creating a new <see cref="BlobServiceClient"/>
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs:20]: Wrong `<see cref...`

File: samples/SubDomain/GetMessage/Program.cs (hunk 3 above)
Comment (FirestarJes) [reply samples/SubDomain/GetMessage/Program.cs]: Fixed

File: source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    /// </summary>
+    public interface IStorageServiceClientFactory
+    {
+        /// <summary>
+        /// Create a new <see cref="BlobServiceClient"/>
```
Comment (michaelmajgaard) [inline source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs:25]: Wrong `<see cref...`

File: source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs (hunk 11 above)
Comment (michaelmajgaard) [reply source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs:20]: Nevermind ...

File: source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs (hunk 12 above)
Comment (michaelmajgaard) [reply source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs:25]: Nevermind

File: source/PostOffice.Communicator.SimpleInjector/source/PostOffice.Communicator.SimpleInjector/ContainerExtensions.cs (hunk 5 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator.SimpleInjector/source/PostOffice.Communicator.SimpleInjector/ContainerExtensions.cs:76]: Fixed

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs (hunk 6 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:44]: Fixed

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs (hunk 7 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:77]: Fixed

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs (hunk 8 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:128]: Fixed

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs (hunk 9 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:90]: fixed

File: source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs (hunk 10 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator.Tests/Storage/StorageHandlerTests.cs:215]: Fixed

File: source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs (hunk 11 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs:20]: That is actually the right cref, as it returns a BlobServiceClient. I named the factory StorageServiceClient to convey that it's just storage, but it doesn't make a descision about which type of storage.

File: source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs (hunk 12 above)
Comment (FirestarJes) [reply source/PostOffice.Communicator/source/PostOffice.Communicator/Factories/IStorageServiceClientFactory.cs:25]: That is actually the right cref, as it returns a BlobServiceClient. I named the factory StorageServiceClient to convey that it's just storage, but it doesn't make a descision about which type of storage.

**Review by michaelmajgaard — APPROVED** [review]

**Review by defectiveAi — APPROVED** [review]

**Pull request merged**